Implement Base64 Encoding, Replacing Deprecated Unescape Method

mohitvdx opened the pull request:
fixes #3585 

The unescape() function is deprecated.

Changes:
The decodeURIComponent() method decodes a URI component used.

@walterbender  please review this pull request

mohitvdx commented:
please review this @walterbender sir

walterbender commented:
Uncaught DOMException: Failed to execute 'btoa' on 'Window': The string to be encoded contains characters outside of the Latin1 range.
    at createHelpContent (file:///home/walter/musicblocks/js/turtledefs.js:480:28)
    at Activity.setupDependencies (file:///home/walter/musicblocks/js/activity.js:302:13)
    at file:///home/walter/musicblocks/js/activity.js:5145:14
    at Object.execCb (file:///home/walter/musicblocks/lib/require.js:113:191)
    at Module.check (file:///home/walter/musicblocks/lib/require.js:67:406)
    at Module.<anonymous> (file:///home/walter/musicblocks/lib/require.js:82:103)
    at file:///home/walter/musicblocks/lib/require.js:22:51
    at file:///home/walter/musicblocks/lib/require.js:84:76
    at each (file:///home/walter/musicblocks/lib/require.js:14:77)
    at Module.emit (file:///home/walter/musicblocks/lib/require.js:84:40)

walterbender commented:
We finally landed something equivalent

mohitvdx commented:
decodeURI() also gives the same error.
I'll look for a better way.

mohitvdx commented:
**Fix for Base64 Encoding of Strings with Non-Latin1 Characters**

This pull request addresses a critical issue encountered when trying to base64 encode strings that contain characters outside of the Latin1 range (ISO-8859-1). Previously, using `window.btoa` directly led to the following error:

```
Uncaught DOMException: Failed to execute 'btoa' on 'Window': The string to be encoded contains characters outside of the Latin1 range.
```

**Problem:**
The `window.btoa` function only supports characters in the Latin1 range. When it encounters characters beyond this range, such as emojis or certain Unicode characters, it fails with a DOMException. This issue was observed in the context of encoding SVG data, which can include a wide range of characters.

**Example:**
Consider a string like `'hello⛳❤️🧀'`. This string combines ASCII characters (`hello`), a character represented by a single 16-bit code unit (`⛳`), and emojis (`❤️` and `🧀`) which are represented by two 16-bit code units or a surrogate pair. Encoding such a string with `btoa` directly results in the aforementioned error.
source: https://web.dev/articles/base64-encoding

**Solution:**
To resolve this, I've implemented a robust encoding and decoding mechanism using `TextEncoder` and `TextDecoder`. 

**Changes Made:**
- Added a new module `base64Utils.js` with two functions: `base64Encode` and `base64Decode`.
- `base64Encode` uses `TextEncoder` to convert a UTF-16 JavaScript string to a UTF-8 encoded Uint8Array. It then converts this Uint8Array into a binary string, which `window.btoa` can safely encode.
- `base64Decode` reverses this process, turning the base64 string back into the original UTF-16 JavaScript string.

**Usage:**
Instead of `window.btoa(unescape(encodeURIComponent(svgData)))`, we now use `window.btoa(base64Encode(svgData))`. This approach is more in line with modern JavaScript practices and avoids the issues with non-Latin1 characters.

This change ensures our base64 encoding is reliable and can handle all types of characters present.

---
Please review this @walterbender sir

walterbender commented:
I am seeing an error.

activity.js:1 Uncaught SyntaxError: Identifier 'base64Encode' has already been declared (at activity.js:1:1)

mohitvdx commented:
I'm unable to find its reason, ```activity.js:1:1``` is a comment. Do you see anything else?

mohitvdx commented:
Hi @walterbender sir, I have fixed the error. 
please review

walterbender commented on js/activity.js:
```diff
@@ -46,7 +46,7 @@
    globalActivity, hideArrows, doAnalyzeProject
  */
-
+//if base64Encode is not defined, define it
```
No reason for this comment.

mohitvdx replied:
sure. Removed it

walterbender commented on js/activity.js:
```diff
@@ -186,6 +186,7 @@ const doAnalyzeProject = function() {
 };
 
 class Activity {
+    
```
please don't add unrelated changes

mohitvdx replied:
yes, fixed it. Thank you for the review.

walterbender commented on js/block.js:
```diff
@@ -1425,7 +1425,7 @@ class Block {
             image.src =
                 "data:image/svg+xml;base64," +
-                window.btoa(unescape(encodeURIComponent(EXPANDBUTTON)));
+                window.btoa(base64Encode(EXPANDUTTON));
```
typo

mohitvdx replied:
fixed it

walterbender commented:
and presumably you tested it?
